### DISCUSSION TITLE
"getStaticProps" like behaviour when using "searchParams"?

### DISCUSSION THREAD

**Discussion by borispoehland** [vercel/next.js#49800] [App Router] [answered]
Hey everyone,

In Next.js 12, this was possible
```
export default function Page() {
  const searchParams = useRouter().query
  // return something 
}

export async function getStaticProps() {
  // do something
}
```

That way the site was statically generated, but during build time the searchParams are `{}` because they cannot be known ahead of time. No problem

In Next.js 13, we want to do the same, but because using `searchParams` or `useSearchParams` automatically opts the page into CSR, how can one use searchParams + have the site statically generated?
```
export default function Page({ searchParams }) {
  // fetch something that should be only fetched on build time and never else
  // return something
}
```
Now, the obvious answer might be: "Well, just use `useSearchParams` + `<Suspense>` so that the CSR doesn't bubble up until the route segment, but is catched earlier by the Suspense boundary. But what if the page itself needs the `searchParams`? With Next.js 12 it was no problem, the page was still generated statically (and with `{}` as `searchParams` during build). But with Next.js 13 it seems no longer possible to say: "Keep the page static and use `searchParams` in them", probably because of the new server rendering environment. 

Am I missing something?

**Comment by icyJoseph** [comment]
> In Next.js 13, we want to do the same, but because using searchParams or useSearchParams automatically opts the page into CSR, how can one use searchParams + have the site statically generated?

I think you mean that `searchParams` at page level, opts into dynamic rendering, and that to use `useSearchParams`, you must use a client component.

With the page directory (what you call next 12), how did you deal with not knowing the query values at build time? You probably used fallbacks, default values, or simply returned null, right? I think the idea is to use suspense boundaries around the parts that must use the search params. 

If the entire page needs the search params to "work", then can we really think of pre-rendering it? All we really do is make an empty skeleton, which needs to be filled-up client side. In other words, in the pages dir, you were able to render a skeleton, and in the app router, the same applies, but you use suspense boundaries to do so.

**Reply by borispoehland** [reply]
Thanks for answering. The challenge I had was this: In the `page`, I needed the `searchParams`, but I didn't want the whole route to opt out of static rendering because the layout should stay static. I managed to deploy such a version by replacing
```
export default function Page({ searchParams }) {
  const onlyFetchOnce = await fetch("https://example.com", { cache: "no-cache" }) // didn't work
  return <ClientComponent searchParams={searchParams} />
}
``` 
with
```
export default function Page() {
 const onlyFetchOnce = await fetch("https://example.com", { cache: "no-cache" })
  return <Suspense>
    <ClientComponentUsingUseSearchParamsInside />
  </Suspense>
}
``` 

That way the route did not completely fall back to CSR, but only essentially the whole page, but at least not the layout (but isn't the layout static always anyways?). You agree with me that this is kinda hacky? But it works at least for the data fetching part!

**Reply by icyJoseph** [reply, marked as answer]
Right, but that's the way to go I believe, for this kind of scenarios. You have a sort of entry point where you can do all of the server side manipulations needed, and then you off-load to client components, which handle interactivity and what not.
